web/packages/teleport/src/DesktopSession/KeyboardHandler.tsx

```diff
@@ -0,0 +1,271 @@
+  }
+
+  /**
+   * Called to finish handling a keyboard event.
+   *
+   * For normal keys, this is called immediately.
+   * For withheld or delayed keys, this is called as the callback when
+   * another key is pressed or released (withheld) or after a delay (delayed).
+   */
+  private finishHandlingKeyboardEvent(
+    cli: TdpClient,
+    e: KeyboardEvent,
+    state: ButtonState
+  ): void {
+    // Release any withheld keys before sending the current key.
+    this.sendWithheldKeys(cli);
+
+    // Special handling for CapsLock on Mac.
+    if (e.code === 'CapsLock' && this.isMac) {
+      // On Mac, every UP or DOWN given to us by the browser corresponds
+      // to a DOWN + UP on the remote machine for CapsLock.
+      cli.sendKeyboardInput('CapsLock', ButtonState.DOWN);
+      cli.sendKeyboardInput('CapsLock', ButtonState.UP);
+    } else {
+      // Otherwise, just pass the event through normally to the server.
+      cli.sendKeyboardInput(e.code, state);
+    }
+
+    // If we're finishing handling for a withheld or delayed key, ensure
+    // that it's cleared from the cache.
+    this.clearWithholdingAndDelay(e.code);
+  }
+
+  /**
+   * This sends all currently withheld keys to the server.
+   */
+  private sendWithheldKeys(cli: TdpClient) {
+    this.withheldDown.forEach((value, code) => {
+      if (value) {
+        cli.sendKeyboardInput(code, ButtonState.DOWN);
+      }
+    });
+
+    this.clearAllWithheldDown();
+  }
+
+  /**
+   * Clears the withheld down cache for all keys.
+   */
+  private clearAllWithheldDown() {
+    this.withheldDown.clear();
+  }
+
+  /**
+   * Clears both caches for a single key.
+   *
+   * Calling this on a key that is not in the cache is a no-op.
+   *
+   * @param code The key code to clear the cache for.
+   */
+  private clearWithholdingAndDelay(code: string) {
+    this.clearWithheldDown(code);
+    this.clearDelayedUp(code);
+  }
+
+  /**
+   * Clears the withheld down cache for a single key.
+   *
+   * Calling this on a key that is not in the cache is a no-op.
+   *
+   * @param code The key code to clear the cache for.
+   */
+  private clearWithheldDown(code: string) {
+    this.withheldDown.delete(code);
+  }
+
+  /**
+   * Clears the delayed up cache for a single key.
+   *
+   * Calling this on a key that is not in the cache is a no-op.
+   *
+   * @param code The key code to clear the cache for.
+   */
+  private clearDelayedUp(code: string) {
+    const timeout = this.delayedUp.get(code);
+    if (timeout !== undefined) {
+      clearTimeout(timeout);
+      this.delayedUp.delete(code);
+    }
+  }
+
+  /**
+   * Called when the canvas loses focus.
+   *
+   * This clears the withheld and delayed keys, so that they are not sent
+   * to the server when the canvas is out of focus.
+   */
+  public onFocusOut() {
+    this.clearAllWithholdingAndDelay();
+    this.syncBeforeNextKey = true;
+  }
+
+  /**
+   * To be called before unmounting the component.
+   */
+  public onUnmount() {
```

Comment:
`KeyboardHandler` seems to be well defined as a sort of class that exists outside of the React environment. To that point, I feel like we don't have to be "leaking" the fact that it ends up being used in a React component. How do you feel about calling this function `dispose` instead of `onUnmount`? This would also look nicer from the perspective of `useEffect` as well, which hides inherently imperative APIs behind a more declarative function call.

Reply:
db77c991e08b3c0b10e4100ddb2a138a02cb46d6

Reply:
db77c991e08b3c0b10e4100ddb2a138a02cb46d6